File: templates/problems/04_stack/05_daily_temperatures/python/solution.py
def solve_daily_temperatures(temperatures: list[int]) -> list[int]:
    """Use a monotonic decreasing stack.

    Time: O(n), Space: O(n)
    """
    n = len(temperatures)
    result = [0] * n
    stack: list[int] = []  # stack of indices
    for i in range(n):
        while stack and temperatures[i] > temperatures[stack[-1]]:
            idx = stack.pop()
            result[idx] = i - idx
        stack.append(i)
    return result

File: templates/problems/04_stack/05_daily_temperatures/typescript/solution.ts
/**
 * Use a monotonic decreasing stack.
 * Time: O(n), Space: O(n)
 */
export function solveDailyTemperatures(temperatures: number[]): number[] {
  const n = temperatures.length;
  const result: number[] = new Array(n).fill(0);
  const stack: number[] = []; // stack of indices
  for (let i = 0; i < n; i++) {
    while (stack.length > 0 && temperatures[i] > temperatures[stack[stack.length - 1]]) {
      const idx = stack.pop()!;
      result[idx] = i - idx;
    }
    stack.push(i);
  }
  return result;
}
